Run and show test results in the editor

jacobmendoza opened the pull request:
Overview: Prototype for integrating a runner of AVA tests in Atom. Not ready to merge, just requesting feedback (https://github.com/sindresorhus/atom-ava/issues/3).

Steps: Link the package. At this stage, there is no menu integration (yet), so pressing ctrl-alt-a is needed to activate the package (right pane will appear). If a test file is active, pressing ctrl-alt-x will trigger the execution of the tests. Every time that one assert is processed, a new line will appear in the UI (the feedback during execution needs to be improved).

About the code: It's just a prototype, so probably there are a lot of things to discuss. Intentionally, there is no effort in this PR for formatted and nice UI. I'm using this [parser](https://github.com/tapjs/tap-parser) for the TAP output, and seems to work very well. Anyway, the idea is to abstract the interaction with it, just in case we could need to replace it (with this code, this is only partially achieved).

<img width="390" alt="Part of the pageres suite" src="https://cloud.githubusercontent.com/assets/2363912/13553871/61873c6c-e38f-11e5-90a1-401bb7031f4e.png">

sindresorhus commented:
Good start! I know we didn't discuss this upfront, but the plugin should use ES2015 (ES6), not CoffeeScript. Nothing against CoffeeScript, but ES2015 is more [future friendly](https://twitter.com/sindresorhus/status/669958889941639168) and better for attracting potential contributors.

> At this stage, there is no menu integration (yet)

What kind of menus were you thinking? I think it's more important to implement command palette commands than menus.

sindresorhus commented on keymaps/ava.cson:
```diff
@@ -0,0 +1,3 @@
+'atom-workspace':
+  'ctrl-alt-a': 'ava:toggle'
+  'ctrl-alt-x': 'ava:run'
```
Is there really any point in having two separate shortcuts? I would think it's enough with just Ctrl+Alt+A which toggles and runs tests.

jacobmendoza replied:
There is no big reason for having two today. I used two because they allowed me to easily test in isolation the two steps. I'll compact them in one. 

If in the future we have a parsing of the test file (something that I'd like to discuss after having a nice stable first version) phase we'll need two. No need to worry about this now.

sindresorhus commented on lib/test-runner-process.coffee:
```diff
@@ -0,0 +1,31 @@
+      folder = filePath.substring(0, filePath.lastIndexOf("/") + 1);
+      file = filePath.split("/").pop()
+
+      command = "ava #{file} --tap"
```
Would be better to use the AVA programmatic API. Example: https://github.com/pine613/fly-ava/blob/master/index.js

jacobmendoza replied:
Of course! Will do!.

jacobmendoza replied:
@sindresorhus, sorry if I'm missing something here.

When trying to use the programmatic API providing an absolute path for running a test I'm getting 'No tests found in ../x/y.js, make sure to import "ava" at the top of your test file'. Same result if I use the CLI in a folder different from the one that hosts the file, as: 
- ava /Users/Me/Folder/test.js --verbose

I have been able to trace the code till test-worker.js but being honest, I don't know why it's failing. Any idea?. Maybe I'm doing something wrong...

Thank you! :)

sindresorhus replied:
AVA 0.13.0 changed the API regarding how files are added. Ensure you pass them to the `.run()` method: https://github.com/sindresorhus/ava/blob/6d5f3225cc712f10e7a0a8a9c6e40d574fd4b251/test/api.js#L23-L27

> Same result if I use the CLI in a folder different from the one that hosts the file, as:

I can reproduce too. Would you mind opening an issue on AVA?

```
~/dev/chalk master
❯ ava /Users/sindresorhus/dev/acosh/test.js

  ✖ No tests found in ../acosh/test.js, make sure to import "ava" at the top of your test file
```

## 

As a dirty workaround, for now you could probably do this https://github.com/sindresorhus/atom-linter-xo/commit/5d85618ad695b9b7142ac4015a6dde6d8709c6f9.

jacobmendoza replied:
No worries, I will open the issue!. I thought that it was more likely to be something wrong with my environment. I saw the change in the API and was providing the files to the run() method. 

I'll take a look at the workaround. Thanks :).

sindresorhus commented on package.json:
```diff
@@ -35,6 +39,10 @@
   ],
+  "dependencies": {
+    "event-kit": "^1.1.0",
+    "tap-parser": ">=v1.2.2"
```
`>=v1.2.2` => `^1.2.2`

jacobmendoza commented:
I wasn't aware of [this](http://blog.atom.io/2015/02/04/built-in-6to5.html) but these are actually good news. No doubt this is more future friendly and has more potential. I'll start doing some testing.

The comment about the menu did not have a big foundation, I was just trying to say that there is no visual entry point in the UI. Probably we don't need it.

jacobmendoza commented:
Thanks for all the feedback :). I'll try to push a new version with these issues addressed asap.

jacobmendoza commented:
I have addressed some of the issues previously marked:

- The files are now ES6.
- Dependencies fixed.
- I'm still not using the programmatic API from AVA. **Pending**.
- Ctrl+A toggles the package and runs the tests.
- I haven't done anything with the UI. At this stage is only showing the tests. There is no method to show errors.

I have tried to implement a non very successful cancellation mechanism. I'm not sure if it's the best approach, but the idea is stop the process if the user closes the package or if the test execution process is launched several times. It'd great have some feedback about the appropriate way to do it with the programmatic API.

Is it looking any better?. Thank you!.

sindresorhus commented on lib/main.js:
```diff
@@ -0,0 +1,34 @@
+'use babel';
```
Use `/** @babel */` instead.

sindresorhus commented on package.json:
```diff
@@ -35,10 +39,15 @@
     "xo": "*"
   },
   "xo": {
+    "envs": ["browser", "node", "jasmine", "atomtest"],
```
One item on a separate line

sindresorhus commented on keymaps/ava.cson:
```diff
@@ -0,0 +1,2 @@
+'atom-workspace':
+  'ctrl-alt-a': 'ava:toggle'
```
Can you use a JSON file here instead?

sindresorhus commented on lib/main.js:
```diff
@@ -0,0 +1,34 @@
+'use babel';
+
+const Panel = require('./panel');
+const {CompositeDisposable} = require('atom');
```
Since it goes through Babel, you can use ES2015 `import` syntax.

sindresorhus commented on package.json:
```diff
@@ -2,6 +2,7 @@
   "name": "ava",
   "version": "0.2.0",
   "description": "Snippets for AVA",
+  "main": "./lib/main",
```
Just rename `main.js` to `index.js` and use `"main": "lib",`

jacobmendoza replied:
Not working for me... not sure if I'm missing something.

sindresorhus commented on spec/main-spec.js:
```diff
@@ -0,0 +1,37 @@
+'use babel';
+
+describe('TestingForAva', () => {
```
Do we have to use Jasmine for testing Atom packages or could we use AVA?

jacobmendoza replied:
Yeah, is more consistent and makes a lot of sense use AVA. I have been thinking about that since the beginning. My only doubt is if it can lead to any problems, just because Jasmine in the de facto option in Atom. 

Probably is time to make some tests.

sindresorhus commented on styles/ava.less:
```diff
@@ -0,0 +1,11 @@
+@import "ui-variables";
```
single-quotes

sindresorhus commented on styles/ava.less:
```diff
@@ -0,0 +1,11 @@
+  .message { padding-bottom: 10px; }
+  .summary { font-size: 12px; padding-top: 10px; }
+  .test { font-size: 11px; font-weight: bold; }
+  .executing { font-size: 12px; font-weight: bold; padding-bottom: 10px;}
```
use tab indentation

sindresorhus commented:
Nice progress :) Added some more feedback.

jacobmendoza commented:
@sindresorhus, just wanted to check if you have any ideas about something that came up while thinking about the plugin and the programmatic api. 

At least speaking about the plugin I think makes sense have a mechanism that allows users to cancel the current execution of the suite. Moreover, in the plugin context, that could control some edge cases, as the user pressing several times the key that runs the tests and things like that.

Any thoughts about that?. Is there any way to get that with the programmatic api?. I understand that it must be difficult due to the async nature of ava, but just in case.

If this makes more sense in gitter or as an issue let me know!.

Thanks,

sindresorhus commented:
> At least speaking about the plugin I think makes sense have a mechanism that allows users to cancel the current execution of the suite.

That makes sense. Not something we currently support though. https://github.com/sindresorhus/ava/issues/710

> Moreover, in the plugin context, that could control some edge cases, as the user pressing several times the key that runs the tests and things like that.

That can easily be worked around for now by not allowing a new run when one is in progress.

-

Feel free to continue using the CLI if you feel it would be better for now.

jacobmendoza commented:
Yeah, after testing a bit I think I'll continue with the CLI (for now) and will apply the workaround as suggested. I think that makes sense.

Progress soon. 

Thanks @sindresorhus!,

jacobmendoza commented:
1. Initial version of the UI. Deliberately, just for now, I have used DOM manipulation. I know that it does not look good. There is an ongoing discussion in GitHub about a view framework for the packages (https://github.com/atom/atom/issues/5756) and I need to take a look.

2. For now we only have the UI being rendered in a panel. When more options appear (block decorations or mini-version in the toolbar, etc.) I think it should be easy to extend after some refactor. For getting something done about that, I will need also a correlation between the result that I get and the line that the test occupies in the file. Will try to get something done about that.

3. Also, for now, it is only executing the current file, not the entire project. I think a decision about a view framework should be made before implementing this.

<img width="337" alt="screen shot 2016-04-30 at 02 20 17" src="https://cloud.githubusercontent.com/assets/2363912/14933051/2c575c78-0e7a-11e6-92c6-cf80c33a01bf.png">

jamestalmage commented:
Oooh! You are tempting me to switch from WebStorm with this!

jacobmendoza commented:
@jamestalmage, lots of things to do yet, some embarrassing enough parts in the code, but hopefully it will get better. I appreciate a lot your comment :).

jacobmendoza commented:
@sindresorhus, @jamestalmage, if you have time it would be great if you can give it a go and let me know what you think:

[Video](https://youtubevideourl)
<img width="200" alt="screen shot 2016-05-02 at 17 32 03" src="https://cloud.githubusercontent.com/assets/2363912/14960304/f73232de-108b-11e6-9bb4-ae7bfc8e9bc4.png">

Some minor changes in the UI to prepare the plugin for executing several files. The idea is that we will have several blocks as the one featured in the video (one per file).

My initial understanding is that some of the features that I feel like missing (block decoration and multiple files) will require some minor modifications in the TAP reporter of AVA, that is why I would like to know if the features that we have make sense as an MVP before going for that second phase.

Thanks!

sindresorhus commented:
@jacobmendoza This looks really great! 🙌 More than good enough for a MVP.

-

Some minor nitpick. When using the Atom Dark theme some things could be tweaked:

<img width="358" alt="screen shot 2016-05-04 at 22 03 11" src="https://cloud.githubusercontent.com/assets/170270/15018531/25b93a62-1244-11e6-85ed-42acd515e39b.png">

Maybe use `rgba()` with lowered opacity on the horizontal line and the `test.js` background. Or maybe there's a built-in theme background variable we could use.

And `✖` instead of the current `x` for closing the sidebar.

-

And I get this when trying to run a test file from the [`got`](https://github.com/sindresorhus/got) project. Maybe because it's in a nested folder? I'm running the command directly from the file.

<img width="356" alt="screen shot 2016-05-04 at 22 10 59" src="https://cloud.githubusercontent.com/assets/170270/15018754/365c2d6a-1245-11e6-9fed-fa52114c7b5a.png">

jacobmendoza commented:
Hey @sindresorhus, thanks for spending some time with this!.

1) I had in mind the style for the dark theme and I forgot in the last minute. Yeah, we should have some variables in place that should help us. I'll take a look.

2) I've changed the close icon already. Will push the commit with all the other changes.

3) That last error that you have there is strange!. I have cloned the project and tried to reproduce. I have opened atom in the project folder, and run the file (gzip.js) with the command and with the keyboard shortcut. Worked for me with both of them, also for the other files of the project. Is it happening just for that one?.

jamestalmage commented:
Very cool.

Only thing I've noticed so far is that skipped tests show as passed.

jfmengels commented:
> Only thing I've noticed so far is that skipped tests show as passed.

And `todo` tests are shown as failed.

Looks like a great start! Great job @jacobmendoza!

jacobmendoza commented:
Hey @jamestalmage, @jfmengels!.

Thanks for trying out the package! :). Yeah, that's important, something funny with the parsing process I'd say. Hopefully it'll be fixed very soon.

jamestalmage commented:
@jacobmendoza 
If it's a parsing problem, that could very well be AVA's fault. Our TAP output isn't super sophisticated. 

https://github.com/jamestalmage/tap-emitter aims to fix that, but I've kind of let that languish. I would love extra help on that. It would require fixing https://github.com/jamestalmage/tap-emitter/issues/2 and integrating into AVA. No pressure, but it might make your life easier to improve the TAP output.

jacobmendoza commented:
@jamestalmage, not sure yet if it's the output that AVA generates or actually the parsing process. I think I'll be able to confirm today.

Anyway, yeah, I would be more than happy to help. This is pretty far from my usual technology stack, so I may need some help for start. Actually, regarding the TAP output, there are a few ideas that I wanted to check with you, but I was trying to get something basic working here before.

I'll keep you posted!

jacobmendoza commented:
@jamestalmage, @jfmengels, thanks for your comments yesterday. They were really helpful.

1) About running tests that were marked as skip/todo, should be working now. AVA is doing just fine, and literally, it wasn't implemented (6d6449a). For now, they just won't appear in the summary and statistics and they will appear in the list in a different colour. If you have an opinion about the presence/shape that they should have in the UI let me know!. Used the most basic approach for now. 

2) @jamestalmage, again, there are more things that I'd like to get from the TAP reporter, but my lack of knowledge is so big that I don't even know if my ideas make sense or what would be the best approach to implement them. I'll contact you after putting this in a reasonable state for a first phase.

3) The plugin should be looking significantly better in different themes. I've tried with 6 and at least the result is reasonable. I'm using some pre-defined variables of the themes (7d3aa5e).

Also, fixed a crash if the user tries to run tests in a window with no active editor.

sindresorhus commented:
> Can I copy it to this package?

Yes, should be in the package.

jacobmendoza commented:
@sindresorhus, did you take a look at the error that you were having while trying to test [got](https://github.com/sindresorhus/got)?. Can you still reproduce it?.

I was wondering if, assuming that the defect is no longer present, we could start thinking about merging with the main branch. This PR is starting to get big, and I would like to make a difference between the first basic features and the followings to come.

Thank you!

sindresorhus commented:
@jacobmendoza Yes, works great now! This is really awesome. We really appreciate all the hard work you put into this.

![superhighfive](https://cloud.githubusercontent.com/assets/170270/15268787/51b60550-1a14-11e6-9a3e-c7929fc4d6fe.gif)

sindresorhus commented:
I'll do a new release and tweet about it next week ;)

jfmengels commented:
![fmkp70c7IIb9C](https://camo.githubusercontent.com/71b3da642b85fccb190cac9b951f6914c153c48e/68747470733a2f2f6d65646961342e67697068792e636f6d2f6d656469612f666d6b703730633749496239432f67697068792e676966>)

jacobmendoza commented:
@sindresorhus, @jfmengels,

Thanks a million for your kind words, really ✋✋✋ . I have enjoyed it a lot and it has been great to have your feedback during the process!.

Planning to comment soon https://github.com/sindresorhus/atom-ava/issues/13 to see what you think.

Cheers guys!
